src/main/java/org/truffleruby/core/array/ArrayWriteNormalizedNode.java

```diff
@@ -72,11 +105,25 @@ public Object writeExtendByOne(DynamicObject array, int index, Object value,
     // Writing beyond the end of an array - may need to generalize to Object[] or otherwise extend
 
     @Specialization(guards = {
-            "!isInBounds(array, index)", "!isExtendingByOne(array, index)", "strategy.matches(array)", "strategy.isPrimitive()"
+            "strategy.isStorageMutable()", "!isInBounds(array, index)", "!isExtendingByOne(array, index)", "strategy.matches(array)", "strategy.isPrimitive()"
     }, limit = "STORAGE_STRATEGIES")
     public Object writeBeyondPrimitive(DynamicObject array, int index, Object value,
             @Cached("of(array)") ArrayStrategy strategy,
             @Cached("create()") ArrayGeneralizeNode generalizeNode) {
+        return writeBeyondPritiveInternal(array, index, value, strategy, generalizeNode);
+    }
+
+    @Specialization(guards = {
+            "!strategy.isStorageMutable()", "!isInBounds(array, index)", "!isExtendingByOne(array, index)", "strategy.matches(array)", "mutableStrategy.isPrimitive()"
+    }, limit = "STORAGE_STRATEGIES")
+    public Object writeBeyondPrimitive(DynamicObject array, int index, Object value,
+            @Cached("of(array)") ArrayStrategy strategy,
+            @Cached("strategy.generalizeForMutation()") ArrayStrategy mutableStrategy,
+            @Cached("create()") ArrayGeneralizeNode generalizeNode) {
+        return writeBeyondPritiveInternal(array, index, value, mutableStrategy, generalizeNode);
+    }
```

Comment:
Can this be merged with the one above by always using `generalizeForMutation()`?